packages/browser/src/customizations/posthogReduxLogger.ts

```diff
@@ -90,68 +93,83 @@ function defaultTitleFunction(stateEvent: StateEvent): string {
 // we need a posthog logger for the rate limiter
 const phConsoleLogger: Logger = createLogger('[PostHog Action RateLimiting]')
 
-function defaultLogger(title: string, stateEvent: StateEvent): void {
+export function browserConsoleLogger(title: string, stateEvent: StateEvent): void {
     // but the posthog logger swallows messages unless debug is on
     // so we don't want to use it in this default logger
     // eslint-disable-next-line no-console
     console.log(title, stateEvent)
 }
 
+/**
+ * Logger that sends state events to PostHog session recordings
+ * Requires that the loaded posthog instance is provided
+ * And returns the function to use as the logger
+ *
+ * e.g. const config = { logger: sessionRecordingLoggerForPostHogInstance(posthog) }
+ */
+export const sessionRecordingLoggerForPostHogInstance: (posthog: PostHog) => PostHogStateLogger =
+    (postHogInstance: PostHog) =>
+    (title: string, stateEvent: StateEvent): void => {
+        postHogInstance?.sessionRecording?.tryAddCustomEvent('app-state', { title, stateEvent })
+    }
+
 /**
  * Get only the changed keys from two states
+ * NB exported for testing purposes only, not part of the public API and may change without warning
+ *
  * Returns { prevState: changedKeysOnly, nextState: changedKeysOnly }
  */
-function getChangedStateKeys<S>(prevState: S, nextState: S): { prevState?: Partial<S>; nextState?: Partial<S> } {
-    if (!isObject(prevState) || !isObject(nextState)) {
-        // we only support objects as state, but can't guarantee that's what we'll get
-        return {}
-    }
+export function getChangedStateKeys<S>(
```

Comment:
this was very AI driven refactoring
but I got the performance and correctness tests in place first

went from >100ms to <30ms for the complex case